### PR TITLE
* bugfix: update mutable state to handle duplicated update activity / user timer for better Cassandra performance

### PR DESCRIPTION
**Pull request by wxing1292**
* bugfix: update mutable state to handle duplicated update activity / user timer for better Cassandra performance

### CODE REVIEW

File: common/persistence/cassandraPersistence.go [hunk 1]
```diff
@@ -2276,10 +2276,10 @@ func (d *cassandraPersistence) updateActivityInfos(batch *gocql.Batch, activityI
 	}
 }
 
-func (d *cassandraPersistence) updateTimerInfos(batch *gocql.Batch, timerInfos []*TimerInfo, deleteInfos []string,
+func (d *cassandraPersistence) updateTimerInfos(batch *gocql.Batch, timerInfos map[*TimerInfo]struct{}, deleteInfos map[string]struct{},
```
Comment (samarabbas) [inline common/persistence/cassandraPersistence.go]: why are you changing this to map?

File: common/persistence/cassandraPersistence.go [hunk 2]
```diff
@@ -2310,10 +2310,10 @@ func (d *cassandraPersistence) updateTimerInfos(batch *gocql.Batch, timerInfos [
 	}
 }
 
-func (d *cassandraPersistence) updateChildExecutionInfos(batch *gocql.Batch, childExecutionInfos []*ChildExecutionInfo,
+func (d *cassandraPersistence) updateChildExecutionInfos(batch *gocql.Batch, childExecutionInfos map[*ChildExecutionInfo]struct{},
```
Comment (samarabbas) [inline common/persistence/cassandraPersistence.go]: looks like you are updating a bunch of parameters which used to be slice to a map.  Is there a good reason, it's not obvious from looking at the change why this is relevant for the bugfix.

File: common/persistence/cassandraPersistence.go [hunk 3]
```diff
@@ -2415,12 +2415,16 @@ func (d *cassandraPersistence) updateSignalInfos(batch *gocql.Batch, signalInfos
 	}
 }
 
-func (d *cassandraPersistence) updateSignalsRequested(batch *gocql.Batch, signalReqIDs []string, deleteSignalReqID string,
+func (d *cassandraPersistence) updateSignalsRequested(batch *gocql.Batch, signalReqIDs map[string]struct{}, deleteSignalReqID string,
 	domainID, workflowID, runID string, condition int64, rangeID int64) {
 
+	signalReqIDsSlice := []string{}
```
Comment (samarabbas) [inline common/persistence/cassandraPersistence.go]: and this is even stranger where you changed the parameter from slice to a map and then convert it back to a slice within the function.

File: service/history/workflowExecutionContext.go [hunk 4]
```diff
@@ -217,6 +217,20 @@ func (c *workflowExecutionContext) updateHelper(builder *historyBuilder, transfe
 		replicationTasks = append(replicationTasks, c.msBuilder.createReplicationTask())
 	}
 
+	// // TODO
```
Comment (samarabbas) [inline service/history/workflowExecutionContext.go]: remove this.

Comment (samarabbas) [reply service/history/workflowExecutionContext.go]: If you want this for debugging then I would recommend to implement String() on mutableStateBuilder.

File: common/persistence/cassandraPersistence.go (hunk 1 above)
Comment (wxing1292) [reply common/persistence/cassandraPersistence.go]: according to what i have seen during debugging, the slice can contains duplicates,

File: common/persistence/cassandraPersistence.go (hunk 2 above)
Comment (wxing1292) [reply common/persistence/cassandraPersistence.go]: "bugfix: update mutable state to handle duplicated update activity / user timer for better Cassandra performance"

File: common/persistence/cassandraPersistence.go (hunk 3 above)
Comment (wxing1292) [reply common/persistence/cassandraPersistence.go]: the database layer is using set, and i cannot simply just use a map here.

File: service/history/workflowExecutionContext.go (hunk 4 above)
Comment (wxing1292) [reply service/history/workflowExecutionContext.go]: good catch.

File: docker/auto-build-docker-image.sh [hunk 5]
```diff
@@ -6,5 +6,5 @@ cm_msg=`git log -1 --abbrev-commit`
 if [ "$last_cm_id" == "$curr_cm_id" ]; then
         echo "Last commit doesn't change."
 else
-        echo "Last commit changed. Start building..."  && docker build . -t ubercadence/server:master --build-arg git_branch=master && docker push ubercadence/server:master &&  mail -s "Docker image auto build succeeded HEAD: $curr_cm_id tag: master" longer@uber.com,maxim@uber.com,yiminc@uber.com,venkat@uber.com,krupapc@uber.com,samar@uber.com,wenquanx@uber.com,boweixu@uber.com,meiliang@uber.com,arthurg@uber.com,nathanbl@uber.com <<< "commit info: $cm_msg  " && echo $curr_cm_id > .last_cm_id.txt && echo "build succ!"
+        echo "Last commit changed. Start building..."  && docker build . -t ubercadence/server:master --build-arg git_branch=master && docker push ubercadence/server:master &&  mail -s "Docker image auto build succeeded HEAD: $curr_cm_id tag: master" longer@uber.com,maxim@uber.com,yiminc@uber.com,venkat@uber.com,mihneao@uber.com,samar@uber.com,madhurav@uber.com,boweixu@uber.com <<< "commit info: $cm_msg  " && echo $curr_cm_id > .last_cm_id.txt && echo "build succ!"
```
Comment (vancexu) [inline docker/auto-build-docker-image.sh]: why change this?

Comment (wxing1292) [reply docker/auto-build-docker-image.sh]: base is not update to date
```
This branch is out-of-date with the base branch
Merge the latest changes from master into this branch.
```

**Pull request merged**